### ISSUE TITLE
fix multi-copy consumers of streams

### ISSUE THREAD

**Issue by afilinski** [Ginko-X/Streaming_NESL#16]
keep track of port number when single consumer uses stream multiple times (e.g, s20 := MapTwo Plus s10 s10)
* list of flags in SState probably should not contain client sids; resposibility of each client to know which flag(s) represent its input streams

**Comment by Ginko-X** [comment]
In the current SVCODE instruction set, as far as I can see, the case where a single consumer uses stream multiple times can only occur in the following instructions/SExps:
```
1. MapTwo OP SId SId   -- for binary operation on integers 

2. InterMergeS [SId]    -- for appending many sequences
3. SegInterS [(SId,SId)] 
4. PriSegInterS [(SId,SId)] 
```
These are binary instructions that allow a variable to be passed as all arguments (`MapTwo` is for operations of two integers, and the last three for (general) `_append` function).
The other instructions, according to their semantics, are impossible.

For `MapTwo`, it is by no means to be possible to get into a deadlock, because it only reads  at most one AVal from each of its channels(ports?) in each round, i.e., there is no `Pin i (\x -> Pin i ...)` in its Proc. So even if the buffer size is 1, it is enough for this Proc to produce new AVals. 
The earlier mentioned deadlock case `let x = 3 in x + x` (in the last week's working progress) actually can be considered a bug in the code.


For the other three instructions of general `_append`, it will enter a deadlock if any two SIds in their argument `[SId]`, or `fst  $ unzip [(SId,SId)]` or `snd $ unzip [(SId,SId)]`  are identical. 


I've tried to solve this problem by replicating the stream so that all the argument SIds are unique. For example, rewrite the code 
```
S1:= ToFlags S0,  
S2:= InterMergeS [S1,S1]
``` 
to 
```
S1:= ToFlags S0, 
S2 := InterMergeS [S1,S3], 
S3 := ToFlags S0`
``` 
Unfortunately, it seems not enough to just copy the stream definition instruction (e.g., copy `S3 := ToFlags S0` from `S1` in the above example) to avoid the deadlock. 
Replicating the definition of the stream at the high-level (e.g., rewrite `let x =&3 in x++x` to `let x=&3, y = &3 in x++y`) should be a correct way, but it usually requires to replicate more than just one instruction at the low level, which seems make the problem more complicated and difficult to control.

**Comment by afilinski** [comment]
Your analysis of when the same stream can appear multiple times as a parameter to the same instruction looks correct. Note, however, that the problem is not so much with translating code like `s++s`, because that's prone to deadlock in SNESL (as opposed to NESL) anyway, since you are effectively traversing the stream `s` twice. Rather, the issue is more with code like `let x = 3 in {x,x}` or `{{x,x} : x in &10}`, which should never deadlock. Also, there may be issues with extracting/printing the final answers from a computation: e.g., `{(x,x) : x in &10}` also seems to deadlock, while `let x=3 in (x,x)` simply crashes with a dataTransBack error.

I think the solution is not to modify the code generator, but rather to change the streaming interpreter to allow the same sid to appear multiple times in the list of consumers of a stream, with each copy having its own flag (or read-counter/cursor in the general case).  Then the buffer can only be switched back to filling mode when all copies of the current buffer contents have been completely read, regardless of whether the consumers are independent instructions, or multiple uses in a single instruction. A simple fix would be to change the identification of a stream consumer from just a SId to a pair (SId, Int), where the Int represents the parameter number ("input port)", as referenced by Pin/rin/rinx in the consumer process.

Another, weakly related issue: maybe you shouldn't use the same type SId for both "source code" stream ids in SExp and "runtime" stream ids in Svctx, because when you add recursion, they will definitely no longer be the same. Maybe for now you can just have a type alias RSId = SId, but so that in the future it will be easy to replace RSId with something else, such as RSId = (Int, SId), corresponding to a unique stack-frame number and an offset within that frame, like in the usual compilation scheme for sequential languages with recursion. (Or ,aybe make the changeover now, to properly disentangle the two uses of SId, but always set the frame number to 0 for now.)

**Comment by Ginko-X** [comment]
> I think the solution is not to modify the code generator, but rather to change the streaming interpreter to allow the same sid to appear multiple times in the list of consumers of a stream, with each copy having its own flag (or read-counter/cursor in the general case)

Yes, modifying the code generator/rewriting the code seems meaningless, and I agree that this solution can solve the problems you mentioned in the first paragraph. 

What I'm working on is to optimize/(almost) reconstruct the `SvcodeProcInterp` module, to solve all the known bugs and part of the problems in #17, making the code more clear and easier to debug and to deal with the future work (i.e., add cost, recursions, etc). The idea is:
-  Add a new layer for the interpreter setting up, including:
    - code rewriting (if necessary)
    - DAG analysis/generation 
    - transforming STree(from SFun and WithCtrl) to SIds to be passed to the interpreter
    - interpreter initialization, i.e., generate the initialized Svctx 

- The reconstructed Svctx will include all the necessary information when a stream/Proc is producing a new value. 
```
type Svctx = [(SId, SState)]
type SState = (BufState, ConsuList, ClientList, Proc ())
type ConsuList = [SId] -- consumer list, to find the consumer Sid when it's `Pin/rin/rinx`
type ClientList  = [((SId,Int), Bool)]  -- outgoing edges, `Int` is something similar to a port number
```

- Then the interpreter's work is just:
    - execute the code/Procs with the initialized Svctx (no need to lookup in Dag, CTable )

At this point, the code (in my local repository) is still messy, and I'm sort of getting stuck and doubt if it's a good idea/necessary to do so.

**Comment by Ginko-X** [comment]
> Another, weakly related issue: maybe you shouldn't use the same type SId for both "source code" stream ids in SExp and "runtime" stream ids in Svctx, because when you add recursion, they will definitely no longer be the same.

This is tricky and doesn't make sense to me by now. Maybe I'll be back here after I fix the known problems in the streaming interpreter.

**Comment by afilinski** [comment]
OK, let's talk about this tomorrow. But my immediate guess is that it shouldn't be necessary to make any large-scale changes (such as introducing a new module) to fix either #16 or the second and third bullets in #17 (I see you already fixed the first one).

**Comment by Ginko-X** [comment]
Yes, if rewriting code is not necessary, then the new module will also be not quite necessary. It's because the setting-up works seem easier to be done in an imperative way than in a functional one, so I thought an independent layer/Monad should be better to do these works.

Thank you for these quick replies.
